zink/codegen/src/event.rs

```diff
@@ -1,19 +1,340 @@
+                    }
+                },
+                [v0, v1, v2, v3] => quote! {
+                    let topic1 = #v0.bytes32();
+                    let topic2 = #v1.bytes32();
+                    let topic3 = #v2.bytes32();
+                    let topic4 = #v3.bytes32();
+                    unsafe {
+                    zink::ffi::evm::log4(stringify!(#variant_name).as_bytes(), topic1, topic2, topic3, topic4);
+                    }
+                },
+                _ => unreachable!(),
+            };
+
+            Ok(quote! {
+                #enum_name::#variant_name(#(#ref_patterns),*) => {
+                    #log_impl
+                }
+            })
+        }
+        Fields::Unit => Ok(quote! {
+            #enum_name::#variant_name => {
+                unsafe {
+                zink::ffi::evm::log0(stringify!(#variant_name).as_bytes())
+                }
+            }
+        }),
+    }
+}
+
+/// Validate field types
+fn validate_types(types: &[&Type]) -> Result<()> {
+    for ty in types {
+        if !is_supported_type(ty) {
+            return Err(Error::new_spanned(
+                ty,
+                format!("Unsupported type for event field: {}", quote!(#ty)),
+            ));
+        }
+    }
+    Ok(())
+}
+
+/// Check if type is supported
+fn is_supported_type(ty: &Type) -> bool {
+    matches!(
+        type_to_string(ty).as_str(),
+        "u8" | "u16"
+            | "u32"
+            | "u64"
+            | "u128"
+            | "i8"
+            | "i16"
+            | "i32"
+            | "i64"
+            | "i128"
+            | "bool"
+            | "String"
+            | "Vec<u8>"
+            | "&str"
+            | "&[u8]"
+            | "[u8;32]"
+            | "Address"
+            | "U256"
+    )
+}
+
+/// Generate ABI signature with validation
+fn generate_abi_signature(
+    enum_name: &syn::Ident,
+    variants: &syn::punctuated::Punctuated<Variant, syn::Token![,]>,
+) -> Result<proc_macro2::TokenStream> {
+    let variant_signatures = variants
+        .iter()
+        .map(|variant| {
+            let variant_name = &variant.ident;
+            let params = match &variant.fields {
+                Fields::Named(fields) => fields
+                    .named
+                    .iter()
+                    .map(|f| validate_and_convert_type(&f.ty))
+                    .collect::<Result<Vec<_>>>()?
+                    .join(","),
+                Fields::Unnamed(fields) => fields
+                    .unnamed
+                    .iter()
+                    .map(|f| validate_and_convert_type(&f.ty))
+                    .collect::<Result<Vec<_>>>()?
+                    .join(","),
+                Fields::Unit => String::new(),
+            };
+
+            Ok(format!("{}({})", variant_name, params))
+        })
+        .collect::<Result<Vec<_>>>()?;
+
+    Ok(quote! {
+        vec![
+            #(#variant_signatures.to_string()),*
+        ].join(";")
+    })
+}
```

Comment:
this has already been implemented here this has already been implemented in https://github.com/zink-lang/zink/blob/main/evm/abi/src/abi.rs#L28 just need to `Abi::parse` in this file